src/validation.cpp

```diff
@@ -3607,6 +3607,7 @@ void PruneStaleBlockIndexes()
     // This isn't going to change while we're processing, so just leave and try again later.
     if (!pindexBestHeader) return;
 
+    LOCK(cs_main);
     // If mapBlockIndex isn't bloated, don't bother taking the time.
     if (chainActive.Height() > static_cast<int>(mapBlockIndex.size() - PRUNE_COUNT)) {
         return;
```

Comment:
This area of code was a major place of inefficiency (primarily later on since IsAncestor took way longer than chainActive.Contains, and it was being done first).  It's probably not something I want to lock through the whole thing, since it's looking for stale tips that are old (`PRUNE_DEPTH`).

I think I would prefer locking to check if mapBlockIndex is bloated; and if so, make a copy, unlock and then let it run with the copy even if mapBlockIndex changes; so we're not locked for the whole check if it's a long check [e.g. on startup].

Reply:
I updated to use a separate lock, but missed this section in the subsequent passes. Will get back to this when I have a chance.

Reply:
Updated to split this in two critical sections, and not use cs_main at all.

Reply:
great work so far!